lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -241,23 +219,124 @@ class LowerXMRPass : public LowerXMRBase<LowerXMRPass> {
+      auto flowAtDst = dataflowAt.find(dst);
+      if (flowAtDst == dataflowAt.end())
+        dataflowAt[dst] = flowAtSrc->getSecond();
+      else {
+        auto oldEntry = flowAtDst->getSecond();
+        if (!refSendList[oldEntry].empty()) {
+          // If a valid dataflow already reaches the destination, emit remark.
+          // This happens when multiple connects attached to an input RefType
+          // port, or multiple instantiation (which is already an error).
+          dst.getDefiningOp()->emitRemark("has multipl reaching definitions");
+        }
+        // Dest already has an entry, make sure to update all the entries in the
+        // dataflow map to point to this new entry.
+        for (auto entr : dataflowAt) {
+          if (entr.getSecond() == oldEntry) {
+            dataflowAt[entr.getFirst()] = flowAtSrc->getSecond();
+          }
+        }
+      }
+    }
+  }
+
+  size_t addReachingSendsEntry(Value atRefVal, Attribute newRef,
+                               Optional<size_t> copyFrom = None) {
+    auto index = refSendList.size();
+    dataflowAt[atRefVal] = index;
+    if (copyFrom) {
+      refSendList.emplace_back(refSendList[copyFrom.value()]);
+      refSendList[index].push_back(newRef);
+    } else if (newRef)
+      refSendList.push_back({newRef});
+    else
+      refSendList.push_back({});
+    return index;
+  }
+
+  void garbageCollect() {
+    // Now erase all the Ops and ports of RefType.
+    // This needs to be done as the last step to ensure uses are erased before
+    // the def is erased.
+    for (auto op : llvm::reverse(opsToRemove)) {
+      op->erase();
+    }
+    for (auto iter : refPortsToRemoveMap)
+      if (auto mod = dyn_cast<FModuleOp>(iter.getFirst()))
+        mod.erasePorts(iter.getSecond());
+      else if (auto inst = dyn_cast<InstanceOp>(iter.getFirst())) {
+        ImplicitLocOpBuilder b(inst.getLoc(), inst);
+        inst.erasePorts(b, iter.getSecond());
+        inst.erase();
+      }
+    opsToRemove.clear();
+    refPortsToRemoveMap.clear();
+    dataflowAt.clear();
+    refSendList.clear();
+  }
+
   /// Cached module namespaces.
   DenseMap<Operation *, ModuleNamespace> moduleNamespaces;
 
-  /// Map of a reference value to the path to the RefSend op. The path to
-  /// RefSend is represented as an ArrayAttr, which is an array of the InnerRef
-  /// to InstanceOps. The path is required since there can be multiple paths to
-  /// the RefSend and we need to identify a unique path. Each ref value can be
-  /// satically resolved to a single remote send op, according to the
-  /// constraints on the RefType. The ArrayAttr ensures that only unique copies
-  /// of the path exist.
-  DenseMap<Value, ArrayAttr> reachingRefSendAt;
+  DenseSet<Operation *> visitedModules;
+  /// Map of a reference value to an entry into refSendList. Each entry in
+  /// refSendList represents the path to RefSend is as an ArrayAttr, which is an
```

Comment:
```suggestion
  /// refSendList represents the path to RefSend as an ArrayAttr, which is an
```
(and re-flow, but if changing off ArrayAttr this'll all be changed anyway...)